services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/modules/outputs/_event_filter.py

```diff
@@ -0,0 +1,209 @@
+                # - event was just added
+                # - already waited more than the wait_interval
+                if (
+                    tracked_event.wait_interval is None
+                    or elapsed_since_detection > tracked_event.wait_interval
+                ):
+                    port_key_dir_path = (
+                        self.outputs_manager.outputs_context.outputs_path / port_key
+                    )
+                    total_wait_for = self.delay_policy.get_wait_interval(
+                        get_directory_total_size(port_key_dir_path)
+                    )
+                    tracked_event.wait_interval = total_wait_for
+
+                # could require to wait more since wait_interval was just updated
+                elapsed_since_detection = current_time - tracked_event.last_detection
+                if elapsed_since_detection < tracked_event.wait_interval:
+                    continue
+
+                # event chain has finished, request event to upload port and
+                # remove event chain from tracking
+                self._upload_events_queue.put_nowait(port_key)
+                self._port_key_tracked_event.pop(port_key, None)
+
+            time.sleep(repeat_interval)
+
+    async def _worker_event_emitter(self) -> None:
+        """checks at fixed intervals if it should emit events to upload"""
+        with ThreadPoolExecutor(max_workers=1) as pool:
+            await asyncio.get_event_loop().run_in_executor(
+                pool, self._worker_blocking_event_emitter
+            )
+
+    async def _worker_upload_events(self) -> None:
+        """enqueues uploads for port  `port_key`"""
+        while True:
+            port_key: Optional[str] = await self._upload_events_queue.get()
+            if port_key is None:
+                break
+
+            logger.debug("Request upload for port_key %s", port_key)
+            await self.outputs_manager.port_key_content_changed(port_key)
+
+    async def enqueue(self, port_key: str) -> None:
+        await self._incoming_events_queue.put(port_key)
+
+    async def start(self) -> None:
+        with log_context(logger, logging.INFO, f"{EventFilter.__name__} start"):
+            self._task_incoming_event_ingestion = create_task(
+                self._worker_incoming_event_ingestion(),
+                name=self._worker_incoming_event_ingestion.__name__,
+            )
+
+            self._keep_event_emitter_running = True
+            self._task_event_emitter = create_task(
+                self._worker_event_emitter(), name=self._worker_event_emitter.__name__
+            )
+
+            self._task_upload_events = create_task(
+                self._worker_upload_events(), name=self._worker_upload_events.__name__
+            )
+
+    async def shutdown(self) -> None:
+        with log_context(logger, logging.INFO, f"{EventFilter.__name__} shutdown"):
+
+            await self._incoming_events_queue.put(None)
```

Comment:
suggestion
```python
tasks_to_stop = [self._task_incoming_event_ingestion, self._task_event_emitter, self._task_upload_events]
for task in tasks_to_stop:
    task.cancel()
await asyncio.gather(*tasks_to_stop, return_exception=True)
```

Reply:
That does not work like that, but I see the point. Tried to make it a bit better.